### PR TITLE
Add hyperlinks and paths validation.

### PR DESCRIPTION
**Pull request by ZePan110**

### CODE REVIEW

**Comment by ZePan110** [comment]
@ctao456  Can you help check  https://github.com/opea-project/docs/blob/a841a9299b5660a271cf5c477cdd66e4e91bdfa6/community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md?plain=1#L8   The tool reports an error when checking, I can't find where this link points to.
image

File: .github/workflows/pr-path-detection.yml
Comment (mkbhanda) [inline .github/workflows/pr-path-detection.yml:124]:
```suggestion
            echo "All relative links valid."
```

File: .github/workflows/pr-path-detection.yml
Comment (mkbhanda) [inline .github/workflows/pr-path-detection.yml:106]:
```suggestion
                        echo "*****Retry successful*****"
```

File: .github/workflows/pr-path-detection.yml
Comment (mkbhanda) [inline .github/workflows/pr-path-detection.yml:97]:
```suggestion
                    echo "*****Retry successful*****"
```

**Comment by ctao456** [comment]
Hi @ZePan110 I submitted a pull request to fix this invalid link.
[https://github.com/opea-project/docs/pull/149](https://github.com/opea-project/docs/pull/149)

File: .github/workflows/pr-path-detection.yml
Comment (mkbhanda) [inline .github/workflows/pr-path-detection.yml:88]: Do we care to distinguish between 200 and others int eh 200 range?
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

Also 300, 400 ranges .. not worth trying again ,, only server not responding worth retries.

**Review by mkbhanda — CHANGES_REQUESTED** [review]

**Comment by ZePan110** [comment]
> Hi @ZePan110 I submitted a pull request to fix this invalid link. #149
Still not accessible. BTW, Should we put such private links?
image
image

**Comment by mkbhanda** [comment]
Private links --- given this is for documentation .. am guessing there should be no private links! Would not help user. The error code should be adequate to prompt them how to fix. Or add a sentence .. perhaps you incorrectly provided a restricted link or something.   I am ambivalent about emails .. what if it is a vulnerability reporting link and it bounces. Put that as issue and lets detail with it in the future. Thank you for this link correctness checker script.

**Comment by ctao456** [comment]
> @ctao456
> 
> > Hi @ZePan110 I submitted a pull request to fix this invalid link. #149
> > Still not accessible. BTW, Should we put such private links?
> > image
> > image

In order for the pr to take effect, I need 2 approvals first.

File: .github/workflows/pr-path-detection.yml
Comment (mkbhanda) [inline .github/workflows/pr-path-detection.yml:40]:
```suggestion
                    echo "*****Retry successful*****"
```

**Review by mkbhanda — COMMENTED** [review]
please address earlier comment about not retrying for certain error codes like 400 range.

File: .github/workflows/pr-path-detection.yml
Comment (ZePan110) [reply .github/workflows/pr-path-detection.yml:94]: This tool simply verifies that the link exists, so judging 200 is the clearest status code to indicate that the request was successful and the resource exists. For other int eh 200 ranges, we have not encountered them.
Also retries are necessary because of redirects.

**Comment by ZePan110** [comment]
> > @ctao456
> > > Hi @ZePan110 I submitted a pull request to fix this invalid link. #149
> > > Still not accessible. BTW, Should we put such private links?
> > > image
> > > image
> 
> In order for the pr to take effect, I need 2 approvals first.
The link doesn't seem to exist
image

**Comment by dbkinder** [comment]
> Hi @ZePan110 I submitted a pull request to fix this invalid link. #149

PR approved and merged, thanks.

**Review by dbkinder — CHANGES_REQUESTED** [review]
Some comments

File: .github/workflows/pr-path-detection.yml
Comment (dbkinder) [inline .github/workflows/pr-path-detection.yml:27]: I was worried that multiple links on one line might get missed if one was a mailto, for example, but I see that you used the -o option that puts each match on a separate line in the output.  Nice.

File: .github/workflows/pr-path-detection.yml
Comment (ZePan110) [reply .github/workflows/pr-path-detection.yml:32]: I agree with you, I haven't had time to contact the owners yet, so I'll have to block them for now.

File: community/rfcs/24-05-16-OPEA-001-Overall-Design.md
Comment (chensuyue) [reply community/rfcs/24-05-16-OPEA-001-Overall-Design.md:41]: What's the difference of adding or not adding this "OPEA Architecture"? It seems the same in the user view.

Comment (ZePan110) [reply community/rfcs/24-05-16-OPEA-001-Overall-Design.md:41]: Actually the content in “” works fine without writing, as other locations can attest. But it will interfere with the tool's recognition of the reference. So I removed it.

File: community/rfcs/24-05-16-OPEA-001-Overall-Design.md
Comment (ZePan110) [reply community/rfcs/24-05-16-OPEA-001-Overall-Design.md:61]: Actually the content in “” works fine without writing, as other locations can attest. But it will interfere with the tool's recognition of the reference. So I removed it.

Comment (chensuyue) [reply community/rfcs/24-05-16-OPEA-001-Overall-Design.md:61]: We have to merge this PR today, let us know if it's valuable to keep the title, we can add back in another PR.

**Review by kevinintel — APPROVED** [review]

**Comment by ZePan110** [comment]
> Some comments

All resolved

**Comment by ZePan110** [comment]
@dbkinder @mkbhanda  Can you merge this PR first?

**Comment by dbkinder** [comment]
We need a way to override the link check failure if you consider the case where there is a circular failure between PRs in two different repos.  We could have the situation where two docs are added in separate repos that link to each other.  Both of these PRs would fail the link check and neither could be merged. :(

I've got a case now where I've got a PR adding a doc in the docs repo and a PR in GenAIExamples that references this new doc that is failing the link check because the PR in the doc repo hasn't been merged yet.

**Comment by dbkinder** [comment]
I see your commits are failing because the linkcheck is reporting an error on links that work just fine if I paste them into a browser:
```
Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/company/opeadev
**********Validation failed, try again**********
Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/dingke/
**********Validation failed, try again**********
Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/malinibhandaru/
**********Validation failed, try again**********
Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/amrhalem/
**********Validation failed, try again**********
Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/roberthafner/
**********Validation failed, try again**********
Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/nathan-cartwright-2008228/
**********Validation failed, try again**********
Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/logan-markewich/
**********Validation failed, try again**********
Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/justincormack/
**********Validation failed, try again**********
Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/melissajmckay/
```
This might also be because some sites report an error when using `curl` or `wget` because they restrict access if it looks like the request is not coming from a browser.  The www.intel.com site reports such an error if it's not a browser accessing the site.  The workaround for that is to add a user-agent string to the command (e.g., `wget  -U "Mozilla/5.0" https://www.intel.com/`).

**Comment by ZePan110** [comment]
> I see your commits are failing because the linkcheck is reporting an error on links that work just fine if I paste them into a browser:
> 
> ```
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/company/opeadev
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/dingke/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/malinibhandaru/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/amrhalem/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/roberthafner/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/nathan-cartwright-2008228/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/logan-markewich/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/justincormack/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/melissajmckay/
> ```
> 
> This might also be because some sites report an error when using `curl` or `wget` because they restrict access if it looks like the request is not coming from a browser. The [www.intel.com](http://www.intel.com) site reports such an error if it's not a browser accessing the site. The workaround for that is to add a user-agent string to the command (e.g., `wget -U "Mozilla/5.0" https://www.intel.com/`).

Thank you for this comment, it is very useful.

**Comment by ZePan110** [comment]
> I see your commits are failing because the linkcheck is reporting an error on links that work just fine if I paste them into a browser:
> 
> ```
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/company/opeadev
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/dingke/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/malinibhandaru/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/amrhalem/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/roberthafner/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/nathan-cartwright-2008228/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/logan-markewich/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/justincormack/
> **********Validation failed, try again**********
> Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/melissajmckay/
> ```
> 
> This might also be because some sites report an error when using `curl` or `wget` because they restrict access if it looks like the request is not coming from a browser. The [www.intel.com](http://www.intel.com) site reports such an error if it's not a browser accessing the site. The workaround for that is to add a user-agent string to the command (e.g., `wget -U "Mozilla/5.0" https://www.intel.com/`).

These are linkedin pages that return status code 999 when accessed using curl, even if the user agent field is set, this means that the request was rejected because a non-human access mode request was detected. 
image

**Comment by mkbhanda** [comment]
> > I see your commits are failing because the linkcheck is reporting an error on links that work just fine if I paste them into a browser:
> > ```
> > Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/company/opeadev
> > **********Validation failed, try again**********
> > Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/dingke/
> > **********Validation failed, try again**********
> > Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/malinibhandaru/
> > **********Validation failed, try again**********
> > Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/amrhalem/
> > **********Validation failed, try again**********
> > Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/roberthafner/
> > **********Validation failed, try again**********
> > Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/nathan-cartwright-2008228/
> > **********Validation failed, try again**********
> > Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/logan-markewich/
> > **********Validation failed, try again**********
> > Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/justincormack/
> > **********Validation failed, try again**********
> > Invalid link from /home/runner/work/docs/docs/README.md: https://www.linkedin.com/in/melissajmckay/
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This might also be because some sites report an error when using `curl` or `wget` because they restrict access if it looks like the request is not coming from a browser. The [www.intel.com](http://www.intel.com) site reports such an error if it's not a browser accessing the site. The workaround for that is to add a user-agent string to the command (e.g., `wget -U "Mozilla/5.0" https://www.intel.com/`).
> 
> Thank you for this comment, it is very useful.

Should one do the wget as an insurance for anything not on github - absolute url or relative. Proactively taking care of some scenarios reduces effort handling exceptions from a performance perspective.

**Comment by dbkinder** [comment]
I found this that might help:

wget follows redirections by default, curl doesn’t. If you add -L, curl works fine.

**Comment by ZePan110** [comment]
> We need a way to override the link check failure if you consider the case where there is a circular failure between PRs in two different repos. We could have the situation where two docs are added in separate repos that link to each other. Both of these PRs would fail the link check and neither could be merged. :(
> 
> I've got a case now where I've got a PR adding a doc in the docs repo and a PR in GenAIExamples that references this new doc that is failing the link check because the PR in the doc repo hasn't been merged yet.

The probability is very small, and there is no way to control the checking behavior between repos for unmerged PRs.

**Comment by ZePan110** [comment]
> I found this that might help:
> 
> wget follows redirections by default, curl doesn’t. If you add -L, curl works fine.

In fact the -L was used in the first validation.

**Comment by dbkinder** [comment]
How does this PR have three approvals from reviewers when it's failing CI checks? @ftian1 @chensuyue @kevinintel

**Comment by ZePan110** [comment]
> How does this PR have three approvals from reviewers when it's failing CI checks? @ftian1 @chensuyue @kevinintel

When the reviewer approved, the CI was executed successfully, but now we see that the reason for the failure is due to an invalid path here. https://github.com/opea-project/docs/blob/45d71e519779426822b2e30cbf916a866f5d3679/scripts/test/test.md?plain=1#L3  I'm not sure if this needs to be fixed or blocked from the detection range?

**Comment by dbkinder** [comment]
I've deleted that test folder. But we should have a way to exclude test issues from your CI checking.

**Comment by ZePan110** [comment]
> I've deleted that test folder. But we should have a way to exclude test issues from your CI checking.

The tool's is to expose invalid links during integration, I don't quite understand what you mean by exclusion.

**Comment by ZePan110** [comment]
Due to the excessive discussion, in order to facilitate review, we will recreate the https://github.com/opea-project/docs/pull/177 
 and continue working on this.

**Pull request closed without merging**
